ccm_web/server/src/invitation/invitation.service.ts

```diff
@@ -0,0 +1,60 @@
+import { Injectable } from '@nestjs/common'
+import { ConfigService } from '@nestjs/config'
+import { Config } from '../config'
+import FormData from 'form-data'
+import { randomUUID } from 'crypto'
+import axios from 'axios'
+import { InvitationAPIError } from './invitation.errors'
+import { CanvasUser } from '../canvas/canvas.interfaces'
+
+// Invitation service based on Cirrus
+
+@Injectable()
+export class InvitationService {
+  url: string
+  entityID: string
+  sponsorName: string
+  key: string
+  secret: string
+
+  constructor (private readonly configService: ConfigService<Config, true>) {
+    const invitationConfig = configService.get('invitation', { infer: true })
+    // const domain = configService.get('server.domain', { infer: true })
+    this.url = invitationConfig.apiURL
+    this.entityID = invitationConfig.apiEntityID
+    this.sponsorName = invitationConfig.apiSponsorName
+    this.key = invitationConfig.apiKey
+    this.secret = invitationConfig.apiSecret
+  }
+
+  async sendInvitations (users: CanvasUser[]): Promise<string> {
+    const userEmails: string[] = users.map(user => user.email)
+    const emailAddressCSV = `emailAddress\n${userEmails.join('\n')}`
+
+    const data = new FormData()
+    data.append('cfile', emailAddressCSV, 'fake_file_name.csv')
+    data.append('spEntityId', this.entityID)
+    data.append('sponsorEppn', this.sponsorName)
+    data.append('clientRequestID', 'ccm-' + randomUUID())
+
+    return axios({
+      method: 'POST',
+      url: this.url,
+      auth: {
+        username: this.key,
+        password: this.secret
+      },
+      headers: { ...data.getHeaders() },
+      data: data
+    })
+      .then((response) => {
+        const results: string = JSON.stringify(response.data)
+        console.log(results)
+        return results
+      })
+      .catch((error) => {
+        console.log(error)
+        throw new InvitationAPIError()
+      })
+  }
```

Comment:
I think we can get this so that it doesn't use the callback style, and instead uses the `try`/`catch` and `async`/`await`. You are also using `axios` directly, instead of the `HttpModule`/`HttpService`, which you are specifying as a dependency though. See this example from `CanvasService`; I think we should stick to one pattern for this, and I personally find the `async`/`await` to be more readable and intuitive.

(Sidebar: The use of `HttpModule` versus a library directly has been one of the more perplexing things about NestJS for me. Maybe I'm just not that familiar with dependency injection and it's benefits though. I guess the rationale is that it makes things easier to test when you don't have a hard dependency (you can mock the service or module, instead of the library internals), but I'm still wrapping my head around it honestly.)

Reply:
Also, prefer `logger.debug` or `logger.info` over `console.log`.

Reply:
Fixed.